src/kernel/norm_fiber/cuda.cu

```diff
@@ -43,54 +43,42 @@ void cuda_kernel(Index m, Index n, Index k, Index batch, Scalar alpha, const T *
  *      norm over slices along the first and the last axes.
  * */
 {
-    Index i2_batched = threadIdx.x + blockIdx.x*blockDim.x;
-    Index i2 = i2_batched % k;
-    Index b = i2_batched / k;
-    Index i0_start = threadIdx.y, i0_step = blockDim.y;
-    Index i1_start = threadIdx.z, i1_step = blockDim.z;
     using Y = typename T::repr_t;
     constexpr Y zero{0.0};
-    // Init sum
-    Y sum = zero;
-    if(b < batch)
+
+    // Calculate global thread ID for batch and k dimensions
+    Index i2_batched = blockIdx.x * blockDim.x + threadIdx.x;
+    Index i2 = i2_batched % k;
+    Index b = i2_batched / k;
+
+    if (b < batch && i2 < k)
     {
-        // Cycle over the third axis of input buffer
-        for(Index i1 = i1_start; i1 < n; i1 += i1_step)
+        Y sum = zero;
+
+        // Compute the hypot over the entire m*n slice for the current k and batch
+        for (Index i1 = 0; i1 < n; ++i1)
         {
-            // Get sum of a corresponding slice
-            const T *src_slice = src + ((i1+b*n)*k+i2)*m;
-            // Cycle over the first axis of input buffer
-            for(Index i0 = i0_start; i0 < m; i0 += i0_step)
+            for (Index i0 = 0; i0 < m; ++i0)
             {
-                // Read value from source and Update sum
-                sum = ::hypot(sum, Y{src_slice[i2*m]});
+                Index src_idx = i0 + i2 * m + i1 * m * k + b * m * k * n;
+                Y val = Y{src[src_idx]};
+                sum = ::hypot(sum, val);
             }
         }
-        __shared__ Y block_sum[1];
-        if(i1_start == 0 and i0_start == 0)
+
+        // Apply alpha scaling and beta scaling, then update dst
+        Y result;
+        if (beta == zero)
         {
-            block_sum[threadIdx.x] = zero;
+            result = Y{::fabs(alpha) * sum};
         }
-        __syncthreads();
-        atomicAdd(&block_sum[threadIdx.x], sum);
-        __syncthreads();
-        // Update output value
-        if(i1_start == 0 and i0_start == 0)
+        else
         {
-            // Save result
-            sum = block_sum[threadIdx.x];
-            if(beta == zero)
-            {
-                sum = Y{::fabs(alpha) * sum};
-            }
-            else
-            {
-                sum = Y{::hypot(beta*Y{dst[i2_batched]}, alpha*sum)};
-            }
-            dst[i2_batched] = Y{sum};
+            Y old_val = Y{dst[i2_batched]};
+            result = Y{::hypot(beta * old_val, alpha * sum)};
```

Comment:
convert `alpha` and `beta` into `Y` type before multplying. Otherwise, if `Scalar` type is `double`, entire `hypot` operation is performed in double precision instead of `Y`. Double precision is bad, because it might be not supported on device.